src/platforms/atomic-kms/server/kms/quirks.cpp

```diff
@@ -56,100 +60,221 @@ class mga::Quirks::Impl
+
+    auto gbm_quirks_for(udev::Device const& device) -> std::shared_ptr<GbmQuirks>
+    {
+        class GbmCreateSurfaceNoFlags : public GbmQuirks::CreateSurfaceFlagsQuirk
+        {
+            auto gbm_create_surface_flags() const -> uint32_t override
+            {
+                return 0;
+            }
+        };
+
+        class GbmSurfaceHasFreeBuffersAlwaysTrue : public GbmQuirks::SurfaceHasFreeBuffersQuirk
+        {
+            auto gbm_surface_has_free_buffers(gbm_surface*) const -> int override
+            {
+                return 1;
+            }
+        };
+
+        class DefaultGbmCreateSurfaceFlags : public GbmQuirks::CreateSurfaceFlagsQuirk
+        {
+            auto gbm_create_surface_flags() const -> uint32_t override
+            {
+                return GBM_BO_USE_RENDERING | GBM_BO_USE_SCANOUT;
+            }
+        };
+
+        class DefaultGbmSurfaceHasFreeBuffers : public GbmQuirks::SurfaceHasFreeBuffersQuirk
+        {
+            auto gbm_surface_has_free_buffers(gbm_surface* gbm_surface) const -> int override
+            {
+                return ::gbm_surface_has_free_buffers(gbm_surface);
+            }
+        };
+
+        auto const driver = mgc::get_device_driver(device.parent().get());
+        auto const devnode = device.devnode();
+        mir::log_debug("Quirks(gbm-create-surface-flags + gbm-surface-has-free-buffers): checking device with devnode: %s, driver %s", devnode, driver);
+
+        auto const devnode_or_driver =
+            [](auto devnode, auto driver, auto devnodes, auto drivers) -> std::optional<std::string>
+        {
+            if (devnodes.contains(devnode))
+                return devnodes.at(devnode);
+
+            if (drivers.contains(driver))
+                return drivers.at(driver);
+
+            return std::nullopt;
+        };
+
+        auto create_surface_impl_name =
+            devnode_or_driver(devnode, driver, gbm_create_surface_flags.devnodes, gbm_create_surface_flags.drivers)
+                .transform(
+                    [](auto impl_name)
+                    {
+                        mir::log_debug(
+                            "Quirks(gbm-create-surface-flags): forcing %s implementation", impl_name.c_str());
+                        return impl_name;
+                    })
+                .or_else(
+                    [&driver] -> std::optional<std::string>
+
+                    {
+                        mir::log_debug(
+                            "Quirks(gbm-create-surface-flags): using default implementation for %s driver", driver);
+                        // Not specified
+                        return driver;
+                    })
+                .value();
+
+        auto create_surface_impl = [&]() -> std::unique_ptr<GbmQuirks::CreateSurfaceFlagsQuirk>
+        {
+            if (create_surface_impl_name == "no-flags" || create_surface_impl_name == "nvidia")
+                return std::make_unique<GbmCreateSurfaceNoFlags>();
```

Comment:
[non-blocking] Having driver defaults as code (as opposed to `mgc::AllowList disable_kms_probe{{"virtio_gpu", "vc4-drm", "v3d"}};` _et alias_ as "data") is probably going to be harder to maintain.

It would also be nice to group the defaults in one place.